Mcowger/virtual provider

mcowger opened the pull request:
## Context

This change introduces a "Virtual Provider" designed to manage API costs and improve reliability by intelligently routing requests across multiple, user-configured AI providers. When a primary provider hits its rate limits (in either requests or tokens), the system automatically falls back to a secondary, and then a backup provider, ensuring service continuity.

## Implementation

The solution is centered around two core components:

**VirtualHandler** : This class acts as a stateful proxy for API requests. It dynamically selects an active provider (primary, secondary, or backup) before each request based on their current usage status. It loads the user's provider profiles and instantiates the necessary API handlers. All outgoing requests are funneled through the VirtualHandler, which delegates the call to the currently active underlying handler.

**UsageTracker** : A singleton class that tracks token and request counts for each provider profile. It persists this usage data in the VS Code global state. The VirtualHandler queries this tracker to determine if a provider is currently within its user-defined rate limits (requests/tokens per minute/hour/day). The tracker prunes data older than 24 hours to prevent unbounded storage growth.

The program flow is as follows:

* A request is made to the Virtual Provider.
* `VirtualHandler.adjustActiveHandler()` is called.
* It checks the primary provider's usage against its limits using `UsageTracker`
* If the primary is over its limit, it checks the secondary provider.
* If both are over their limits, it defaults to the backup provider.
* The request is then sent to whichever provider was selected as active.
* The response stream is intercepted to count token usage, which is recorded by the UsageTracker.
* This implementation stores usage data locally in the user's global state, which avoids the need for a separate database



## Screenshots

<img width="387" height="110" alt="Screenshot 2025-07-23 at 12 02 19 PM" src="https://github.com/user-attachments/assets/b66efb78-92d3-49fa-8bd3-76eade33b308" />


<img width="509" height="149" alt="Screenshot 2025-07-23 at 12 01 48 PM" src="https://github.com/user-attachments/assets/50f40791-f429-4aaf-a5d8-edd9ef050f23" />

<img width="387" height="110" alt="Screenshot 2025-07-23 at 12 02 19 PM" src="https://github.com/user-attachments/assets/f6ee31bd-d74e-4f2d-a8ab-e59e29ae64e1" />

<img width="362" height="439" alt="Screenshot 2025-07-23 at 12 02 40 PM" src="https://github.com/user-attachments/assets/a94d6c6f-b8c4-4790-aed6-8a513dbe0c89" />

<img width="382" height="708" alt="Screenshot 2025-07-23 at 12 03 32 PM" src="https://github.com/user-attachments/assets/bf4139a4-48ee-4339-84ee-be5145fff5a6" />

## How to Test

* Enable the Virtual Provider experiments flag in settings
* Create at least 3 separate, non-virtual provider profiles
* Create a new provider profile and set its "API Provider" to "Virtual".
* In the Virtual provider settings:
* Select a profile as the Primary Provider.
* Set a very low rate limit, e.g., 1 Request per minute.
* Select your second profile as the Secondary Provider.
* Select your third profile as the Backup Provider.
* Set the Virtual provider as your active profile.
* Make one request. It should succeed using the Primary provider (OpenAI).
* Immediately make a second request. The system should detect the rate limit on the primary provider and automatically switch to the Secondary provider for this request.

## Get in Touch

mcowger on Discord

mcowgeratsigma committed:
Initial selection and settings.

mcowgeratsigma committed:
Add usage tracker class, getModel and createMessage passthroughs

mcowgeratsigma committed:
Fix uneeded import

mcowgeratsigma committed:
Confirmed changing active handler.

mcowgeratsigma committed:
consumption tracking

mcowgeratsigma committed:
removing debugging statements.

mcowgeratsigma committed:
Working rollovers.

mcowgeratsigma committed:
Fix unused import

mcowgeratsigma committed:
Add use of feature flag

mcowgeratsigma committed:
Unused import

mcowgeratsigma committed:
Write and commit tests

mcowgeratsigma committed:
Update extension state test.

mcowgeratsigma committed:
Add changeset

Commit:
refactor: pre-code review cleanup for virtual provider

- Fix typo in provider-settings.ts comment (defintition -> definition)
- Remove debug console logging from virtual.ts (11 statements)
- Replace 'any' type with proper 'Event' type in Virtual.tsx
- Improve type safety and code cleanliness

hassoncs commented:
I chatted with @mcowger on discord! I'm going to make a few changes and then get this branch merged tomorrow-
* Use N providers instead of primary / secondary / backup
* Translations / Docs
* Testing pass

Commit:
refactor: remove virtualQuotaFallbackProvider experiment

The virtualQuotaFallbackProvider experiment has been removed. The functionality is now always available and no longer controlled by an experiment flag.

This commit removes all references to `virtualQuotaFallbackProvider` from experiment definitions, configurations, and related tests. The `ApiOptions` component in the webview UI has been updated to reflect that the virtual quota fallback provider is always available.

Commit:
Final cleanup changes

Commit:
refactor(virtual-quota-fallback): Rename providers to profiles

Renames all instances of "provider" to "profile" within the virtual quota fallback system. This includes:
- Renaming `providers` array to `profiles` in configuration and types.
- Updating variable names like `providerId` to `profileId` and `providerLimits` to `profileLimits`.
- Adjusting related test cases and internal logic to reflect the new terminology.

This change improves clarity and consistency by aligning the terminology with the concept of managing API profiles rather than generic providers.

Commit:
refactor(ui): streamline profile manipulation in virtual quota fallback

Refactor `VirtualQuotaFallbackProvider` to introduce a `swapProfiles` utility function. This change simplifies the `moveProfileUp` and `moveProfileDown` callbacks by abstracting the profile array manipulation logic. Also, removed redundant comments to improve code readability.

Commit:
final

Commit:
Fix strings provider -> profile

hassoncs reviewed (approved):
Awesome work!! Thanks again! Sending you some free credits now!

deyil commented:
It would be cool if the virtual provider could also support fallbacks and handle API errors. This means that if quota limits are not set, the virtual provider's fallback function could still work automatically during an API error. There could be a setting in each provider specifying after how many errors the fallback to the next provider could activate.
